VisualPinball.Unity/VisualPinball.Unity.Editor/Utils/DataFieldUtils.cs

```diff
@@ -0,0 +1,226 @@
+			if (EditorGUI.EndChangeCheck())
+			{
+				string message = "";
+				List<UnityEngine.Object> recordObjs = new List<UnityEngine.Object>();
+				if (onChange(label, out message, recordObjs, pList))
+				{
+					Undo.RecordObjects(recordObjs.ToArray(), $"{message}");
+				}
+				field = val;
+				return true;
+			}
+			return false;
+		}
+
+		public static bool ItemDataField(string label, ref bool field, OnEditionChangeDelegate onChange, params (string, object)[] pList)
+		{
+			EditorGUI.BeginChangeCheck();
+			bool val = EditorGUILayout.Toggle(label, field);
+			if (EditorGUI.EndChangeCheck())
+			{
+				string message = "";
+				List<UnityEngine.Object> recordObjs = new List<UnityEngine.Object>();
+				if (onChange(label, out message, recordObjs, pList))
+				{
+					Undo.RecordObjects(recordObjs.ToArray(), $"{message}");
+				}
+				field = val;
+				return true;
+			}
+			return false;
+		}
+
+		public static bool ItemDataField(string label, ref Vertex2D field, OnEditionChangeDelegate onChange, params (string, object)[] pList)
+		{
+			EditorGUI.BeginChangeCheck();
+			Vertex2D val = EditorGUILayout.Vector2Field(label, field.ToUnityVector2()).ToVertex2D();
+			if (EditorGUI.EndChangeCheck())
+			{
+				string message = "";
+				List<UnityEngine.Object> recordObjs = new List<UnityEngine.Object>();
+				if (onChange(label, out message, recordObjs, pList))
+				{
+					Undo.RecordObjects(recordObjs.ToArray(), $"{message}");
+				}
+				field = val;
+				return true;
+			}
+			return false;
+		}
+
+		public static bool ItemDataField(string label, ref Vertex3D field, OnEditionChangeDelegate onChange, params (string, object)[] pList)
+		{
+			EditorGUI.BeginChangeCheck();
+			Vertex3D val = EditorGUILayout.Vector3Field(label, field.ToUnityVector3()).ToVertex3D();
+			if (EditorGUI.EndChangeCheck())
+			{
+				string message = "";
+				List<UnityEngine.Object> recordObjs = new List<UnityEngine.Object>();
+				if (onChange(label, out message, recordObjs, pList))
+				{
+					Undo.RecordObjects(recordObjs.ToArray(), $"{message}");
+				}
+				field = val;
+				return true;
+			}
+			return false;
+		}
+
+		public static bool ItemDataField(string label, ref Engine.Math.Color field, OnEditionChangeDelegate onChange, params (string, object)[] pList)
+		{
+			EditorGUI.BeginChangeCheck();
+			Engine.Math.Color val = EditorGUILayout.ColorField(label, field.ToUnityColor()).ToEngineColor();
+			if (EditorGUI.EndChangeCheck())
+			{
+				string message = "";
+				List<UnityEngine.Object> recordObjs = new List<UnityEngine.Object>();
+				if (onChange(label, out message, recordObjs, pList))
+				{
+					Undo.RecordObjects(recordObjs.ToArray(), $"{message}");
+				}
+				field = val;
+				return true;
+			}
+			return false;
+		}
+
+		public static UnityEngine.Object ItemObjectField(string label, UnityEngine.Object field, bool allowSceneObject, OnEditionChangeDelegate onChange, params (string, object)[] pList)

```

Comment:
we need to still pass in the explicit type here, since the "field" var could be null

Reply:
I would like to get rid of those `params (string, object)[] pList` altogether. @Vroonsh it would be great to have more insights of this. Is there an alternative? How many different types are we talking about that made you passing around `object`s?